presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/ArrowPageSource.java

```diff
@@ -0,0 +1,589 @@
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.StandardErrorCode;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import org.apache.arrow.flight.FlightRuntimeException;
+import org.apache.arrow.flight.FlightStream;
+import org.apache.arrow.flight.Ticket;
+import org.apache.arrow.vector.BigIntVector;
+import org.apache.arrow.vector.BitVector;
+import org.apache.arrow.vector.DateDayVector;
+import org.apache.arrow.vector.DateMilliVector;
+import org.apache.arrow.vector.DecimalVector;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.Float4Vector;
+import org.apache.arrow.vector.Float8Vector;
+import org.apache.arrow.vector.IntVector;
+import org.apache.arrow.vector.NullVector;
+import org.apache.arrow.vector.SmallIntVector;
+import org.apache.arrow.vector.TimeMicroVector;
+import org.apache.arrow.vector.TimeMilliVector;
+import org.apache.arrow.vector.TimeSecVector;
+import org.apache.arrow.vector.TimeStampMicroVector;
+import org.apache.arrow.vector.TimeStampMilliVector;
+import org.apache.arrow.vector.TimeStampSecVector;
+import org.apache.arrow.vector.TinyIntVector;
+import org.apache.arrow.vector.VarBinaryVector;
+import org.apache.arrow.vector.VarCharVector;
+import org.apache.arrow.vector.VectorSchemaRoot;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.plugin.arrow.ArrowErrorCode.ARROW_FLIGHT_ERROR;
+
+public class ArrowPageSource
+        implements ConnectorPageSource
+{
+    private final ArrowSplit split;
+    private final List<ArrowColumnHandle> columnHandles;
+    private boolean completed;
+    private int currentPosition;
+    private ArrowFlightClient flightClient;
+    private FlightStream flightStream;
+    private Logger logger = Logger.get(ArrowPageSource.class);
+    private final ArrowFlightClientHandler clientHandler;
+    private Optional<VectorSchemaRoot> vectorSchemaRoot = Optional.empty();
+
+    public ArrowPageSource(ArrowSplit split, List<ArrowColumnHandle> columnHandles, ArrowFlightClientHandler clientHandler,
+            ConnectorSession connectorSession)
+    {
+        this.columnHandles = columnHandles;
+        this.split = split;
+        this.clientHandler = clientHandler;
+        getFlightStream(clientHandler, split.getTicket(), connectorSession);
+    }
+
+    private void getFlightStream(ArrowFlightClientHandler clientHandler, byte[] ticket, ConnectorSession connectorSession)
+    {
+        try {
+            Optional<String> uri = split.getLocationUrls().isEmpty() ? Optional.empty() : Optional.of(split.getLocationUrls().get(0));
+            flightClient = clientHandler.getClient(uri);
+            // Use the token that was sent in the split instead of generating a new one.
```

Comment:
We need to use the plugin that sets extra credentials instead of getting token from split. We will merge that change once the plugin is ready.

Reply:
okay @elbinpallimalilibm